File: krishideep_app/lib/models/market_price.dart
class MarketPrice {
  final String id;
  final String cropName;
  final String cropNameHindi;
  final String variety;
  final String varietyHindi;
  final double currentPrice; // per quintal in INR
  final double previousPrice;
  final double changePercentage;
  final String season; // Kharif, Rabi, Commercial
  final String seasonHindi;
  final DateTime timestamp;
  final String priceUnit;
  final String grade;
  final bool isMSP; // Whether this is MSP or market price
  final double msp202425;
  final double msp202324;
  final double mspIncrease;
  final double mspIncreasePercentage;

  MarketPrice({
    required this.id,
    required this.cropName,
    required this.cropNameHindi,
    required this.variety,
    required this.varietyHindi,
    required this.currentPrice,
    required this.previousPrice,
    required this.changePercentage,
    required this.season,
    required this.seasonHindi,
    required this.timestamp,
    this.priceUnit = 'per quintal',
    this.grade = 'FAQ',
    this.isMSP = false,
    this.msp202425 = 0.0,
    this.msp202324 = 0.0,
    this.mspIncrease = 0.0,
    this.mspIncreasePercentage = 0.0,
  });

  bool get isPriceIncreased => changePercentage > 0;
  bool get isPriceDecreased => changePercentage < 0;

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'cropName': cropName,
      'cropNameHindi': cropNameHindi,
      'variety': variety,
      'varietyHindi': varietyHindi,
      'currentPrice': currentPrice,
      'previousPrice': previousPrice,
      'changePercentage': changePercentage,
      'season': season,
      'seasonHindi': seasonHindi,
      'timestamp': timestamp.toIso8601String(),
      'priceUnit': priceUnit,
      'grade': grade,
      'isMSP': isMSP,
      'msp202425': msp202425,
      'msp202324': msp202324,
      'mspIncrease': mspIncrease,
      'mspIncreasePercentage': mspIncreasePercentage,
    };
  }

  factory MarketPrice.fromJson(Map<String, dynamic> json) {
    return MarketPrice(
      id: json['id'],
      cropName: json['cropName'],
      cropNameHindi: json['cropNameHindi'],
      variety: json['variety'] ?? '',
      varietyHindi: json['varietyHindi'] ?? '',
      currentPrice: json['currentPrice'],
      previousPrice: json['previousPrice'],
      changePercentage: json['changePercentage'],
      season: json['season'] ?? '',
      seasonHindi: json['seasonHindi'] ?? '',
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'])
          : DateTime.now(),
      priceUnit: json['priceUnit'] ?? 'per quintal',
      grade: json['grade'] ?? 'FAQ',
      isMSP: json['isMSP'] ?? false,
      msp202425: json['msp202425']?.toDouble() ?? 0.0,
      msp202324: json['msp202324']?.toDouble() ?? 0.0,
      mspIncrease: json['mspIncrease']?.toDouble() ?? 0.0,
      mspIncreasePercentage: json['mspIncreasePercentage']?.toDouble() ?? 0.0,
    );
  }
}

class PriceTrend {
  final String cropName;
  final List<PricePoint> priceHistory;
  final double averagePrice;
  final double highestPrice;
  final double lowestPrice;
  final String trend; // increasing, decreasing, stable
  final double volatilityIndex;

  PriceTrend({
    required this.cropName,
    required this.priceHistory,
    required this.averagePrice,
    required this.highestPrice,
    required this.lowestPrice,
    required this.trend,
    required this.volatilityIndex,
  });

  Map<String, dynamic> toJson() {
    return {
      'cropName': cropName,
      'priceHistory': priceHistory.map((p) => p.toJson()).toList(),
      'averagePrice': averagePrice,
      'highestPrice': highestPrice,
      'lowestPrice': lowestPrice,
      'trend': trend,
      'volatilityIndex': volatilityIndex,
    };
  }

  factory PriceTrend.fromJson(Map<String, dynamic> json) {
    return PriceTrend(
      cropName: json['cropName'],
      priceHistory: (json['priceHistory'] as List)
          .map((p) => PricePoint.fromJson(p))
          .toList(),
      averagePrice: json['averagePrice'],
      highestPrice: json['highestPrice'],
      lowestPrice: json['lowestPrice'],
      trend: json['trend'],
      volatilityIndex: json['volatilityIndex'],
    );
  }
}

class PricePoint {
  final DateTime date;
  final double price;

  PricePoint({
    required this.date,
    required this.price,
  });

  Map<String, dynamic> toJson() {
    return {
      'date': date.toIso8601String(),
      'price': price,
    };
  }

  factory PricePoint.fromJson(Map<String, dynamic> json) {
    return PricePoint(
      date:
          json['date'] != null ? DateTime.parse(json['date']) : DateTime.now(),
      price: json['price'],
    );
  }
}

class MarketAlert {
  final String id;
  final String cropName;
  final String alertType; // price_drop, price_surge, best_time_to_sell
  final String title;
  final String titleHindi;
  final String message;
  final String messageHindi;
  final double targetPrice;
  final double currentPrice;
  final DateTime createdAt;
  final bool isActive;

  MarketAlert({
    required this.id,
    required this.cropName,
    required this.alertType,
    required this.title,
    required this.titleHindi,
    required this.message,
    required this.messageHindi,
    required this.targetPrice,
    required this.currentPrice,
    required this.createdAt,
    this.isActive = true,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'cropName': cropName,
      'alertType': alertType,
      'title': title,
      'titleHindi': titleHindi,
      'message': message,
      'messageHindi': messageHindi,
      'targetPrice': targetPrice,
      'currentPrice': currentPrice,
      'createdAt': createdAt.toIso8601String(),
      'isActive': isActive,
    };
  }

  factory MarketAlert.fromJson(Map<String, dynamic> json) {
    return MarketAlert(
      id: json['id'],
      cropName: json['cropName'],
      alertType: json['alertType'],
      title: json['title'],
      titleHindi: json['titleHindi'],
      message: json['message'],
      messageHindi: json['messageHindi'],
      targetPrice: json['targetPrice'],
      currentPrice: json['currentPrice'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      isActive: json['isActive'] ?? true,
    );
  }
}

class BestSellingLocation {
  final String marketName;
  final String marketNameHindi;
  final String state;
  final String stateHindi;
  final double price;
  final double distance; // in km
  final String transportCost;
  final double netPrice;
  final String contact;
  final double rating;

  BestSellingLocation({
    required this.marketName,
    required this.marketNameHindi,
    required this.state,
    required this.stateHindi,
    required this.price,
    required this.distance,
    required this.transportCost,
    required this.netPrice,
    required this.contact,
    required this.rating,
  });

  Map<String, dynamic> toJson() {
    return {
      'marketName': marketName,
      'marketNameHindi': marketNameHindi,
      'state': state,
      'stateHindi': stateHindi,
      'price': price,
      'distance': distance,
      'transportCost': transportCost,
      'netPrice': netPrice,
      'contact': contact,
      'rating': rating,
    };
  }

  factory BestSellingLocation.fromJson(Map<String, dynamic> json) {
    return BestSellingLocation(
      marketName: json['marketName'],
      marketNameHindi: json['marketNameHindi'],
      state: json['state'],
      stateHindi: json['stateHindi'],
      price: json['price'],
      distance: json['distance'],
      transportCost: json['transportCost'],
      netPrice: json['netPrice'],
      contact: json['contact'],
      rating: json['rating'],
    );
  }
}

File: krisideepweb/src/models/MarketPrice.ts
export interface MarketPrice {
  id: string;
  cropName: string;
  cropNameHindi: string;
  currentPrice: number; // per quintal in INR
  previousPrice: number;
  changePercentage: number;
  market: string;
  marketHindi: string;
  state: string;
  stateHindi: string;
  timestamp: Date;
  priceUnit: string;
  grade: string;
}

export interface PriceTrend {
  cropName: string;
  priceHistory: PricePoint[];
  averagePrice: number;
  highestPrice: number;
  lowestPrice: number;
  trend: string; // increasing, decreasing, stable
  volatilityIndex: number;
}

export interface PricePoint {
  date: Date;
  price: number;
}

export interface MarketAlert {
  id: string;
  cropName: string;
  alertType: string; // price_drop, price_surge, best_time_to_sell
  title: string;
  titleHindi: string;
  message: string;
  messageHindi: string;
  targetPrice: number;
  currentPrice: number;
  createdAt: Date;
  isActive: boolean;
}

export interface BestSellingLocation {
  marketName: string;
  marketNameHindi: string;
  state: string;
  stateHindi: string;
  price: number;
  distance: number; // in km
  transportCost: string;
  netPrice: number;
  contact: string;
  rating: number;
}

export interface MarketPriceJson {
  id: string;
  cropName: string;
  cropNameHindi: string;
  currentPrice: number;
  previousPrice: number;
  changePercentage: number;
  market: string;
  marketHindi: string;
  state: string;
  stateHindi: string;
  timestamp: string;
  priceUnit: string;
  grade: string;
}

export interface PriceTrendJson {
  cropName: string;
  priceHistory: any[];
  averagePrice: number;
  highestPrice: number;
  lowestPrice: number;
  trend: string;
  volatilityIndex: number;
}

export interface PricePointJson {
  date: string;
  price: number;
}

export interface MarketAlertJson {
  id: string;
  cropName: string;
  alertType: string;
  title: string;
  titleHindi: string;
  message: string;
  messageHindi: string;
  targetPrice: number;
  currentPrice: number;
  createdAt: string;
  isActive: boolean;
}

export interface BestSellingLocationJson {
  marketName: string;
  marketNameHindi: string;
  state: string;
  stateHindi: string;
  price: number;
  distance: number;
  transportCost: string;
  netPrice: number;
  contact: string;
  rating: number;
}

export const marketPriceFromJson = (json: MarketPriceJson): MarketPrice => ({
  id: json.id,
  cropName: json.cropName,
  cropNameHindi: json.cropNameHindi,
  currentPrice: json.currentPrice,
  previousPrice: json.previousPrice,
  changePercentage: json.changePercentage,
  market: json.market,
  marketHindi: json.marketHindi,
  state: json.state,
  stateHindi: json.stateHindi,
  timestamp: new Date(json.timestamp),
  priceUnit: json.priceUnit || 'per quintal',
  grade: json.grade || 'FAQ',
});

export const priceTrendFromJson = (json: PriceTrendJson): PriceTrend => ({
  cropName: json.cropName,
  priceHistory: json.priceHistory.map((p: any) => pricePointFromJson(p)),
  averagePrice: json.averagePrice,
  highestPrice: json.highestPrice,
  lowestPrice: json.lowestPrice,
  trend: json.trend,
  volatilityIndex: json.volatilityIndex,
});

export const pricePointFromJson = (json: PricePointJson): PricePoint => ({
  date: new Date(json.date),
  price: json.price,
});

export const marketAlertFromJson = (json: MarketAlertJson): MarketAlert => ({
  id: json.id,
  cropName: json.cropName,
  alertType: json.alertType,
  title: json.title,
  titleHindi: json.titleHindi,
  message: json.message,
  messageHindi: json.messageHindi,
  targetPrice: json.targetPrice,
  currentPrice: json.currentPrice,
  createdAt: new Date(json.createdAt),
  isActive: json.isActive ?? true,
});

export const bestSellingLocationFromJson = (json: BestSellingLocationJson): BestSellingLocation => ({
  marketName: json.marketName,
  marketNameHindi: json.marketNameHindi,
  state: json.state,
  stateHindi: json.stateHindi,
  price: json.price,
  distance: json.distance,
  transportCost: json.transportCost,
  netPrice: json.netPrice,
  contact: json.contact,
  rating: json.rating,
});

export const marketPriceToJson = (price: MarketPrice): MarketPriceJson => ({
  id: price.id,
  cropName: price.cropName,
  cropNameHindi: price.cropNameHindi,
  currentPrice: price.currentPrice,
  previousPrice: price.previousPrice,
  changePercentage: price.changePercentage,
  market: price.market,
  marketHindi: price.marketHindi,
  state: price.state,
  stateHindi: price.stateHindi,
  timestamp: price.timestamp.toISOString(),
  priceUnit: price.priceUnit,
  grade: price.grade,
});

export const priceTrendToJson = (trend: PriceTrend): PriceTrendJson => ({
  cropName: trend.cropName,
  priceHistory: trend.priceHistory.map(pricePointToJson),
  averagePrice: trend.averagePrice,
  highestPrice: trend.highestPrice,
  lowestPrice: trend.lowestPrice,
  trend: trend.trend,
  volatilityIndex: trend.volatilityIndex,
});

export const pricePointToJson = (point: PricePoint): PricePointJson => ({
  date: point.date.toISOString(),
  price: point.price,
});

export const marketAlertToJson = (alert: MarketAlert): MarketAlertJson => ({
  id: alert.id,
  cropName: alert.cropName,
  alertType: alert.alertType,
  title: alert.title,
  titleHindi: alert.titleHindi,
  message: alert.message,
  messageHindi: alert.messageHindi,
  targetPrice: alert.targetPrice,
  currentPrice: alert.currentPrice,
  createdAt: alert.createdAt.toISOString(),
  isActive: alert.isActive,
});

export const bestSellingLocationToJson = (location: BestSellingLocation): BestSellingLocationJson => ({
  marketName: location.marketName,
  marketNameHindi: location.marketNameHindi,
  state: location.state,
  stateHindi: location.stateHindi,
  price: location.price,
  distance: location.distance,
  transportCost: location.transportCost,
  netPrice: location.netPrice,
  contact: location.contact,
  rating: location.rating,
});
